BITSWAP.md

```diff
@@ -79,122 +75,126 @@ message Message {
-3.  close bitswap stream
+### Interaction Pattern
 
-Events:
+Given that a client C wants to fetch data from some server S:
 
-  - bitswap.addedBlock(block)
-      - see if any peers want this block, and send it
-  - bitswap.getBlock(key, cb)
-      - add to wantlist
-      - maybe send wantlist updates to peers
-  - bitswap.cancelGet(key)
-      - so that can send wantlist cancels
-  - bitswap.receivedMessage(msg)
-      - process the wantlist changes
-      - process the blocks
-  - bitswap.peerConnected(peer)
-      - add peer to peer set + send them wantlist (maybe)
-  - bitswap.peerDisconnected(peer)
-      - remove peer from peer set
+1. C opens a stream `s_want` to S and sends a message for the blocks it wants
+    1. C may either send a complete wantlist, or an update to an outstanding wantlist
+    2. C may reuse this stream to send new wants
+    3. For each of the items in the wantlist C may ask if S has the block (i.e. a Have request) or for S to send the block (i.e. a block request). C may also ask S to send back a DontHave message in the event it doesn't have the block
+2. S responds back on a stream `s_receive`. S may reuse this stream to send back subsequent responses
+    1. If C sends S a Have request for data S has (and is willing to give to C) it should respond with a Have, although it may instead respond with the block itself (e.g. if the block is very small)
+    2. If C sends S a Have request for data S does not have (or has but is not willing to give to C) and C has requested for DontHave responses then S should respond with DontHave
+    3. S may choose to include the number of bytes that are pending to be sent to C in the response message
+3. When C no longer needs a block it previously asked for it should send a Cancel message for that request to any peers that have not already responded about that particular block. It should particularly send Cancel messages for Block requests (as opposed to Have requests) that have not yet been answered.
 
-Tricky Bits:
+### Wire Format
 
-  - clients to bitswap may call `getBlock` then `cancelBlock`
-  - partners may spam wantlists
-  - normalize priorities only per-peer
+```
+message Message {
 
-Modules:
+  message Wantlist {
+    enum WantType {
+      Block = 0;
+      Have = 1;
+    }
 
-  - bitswap-decision-engine
-  - bitswap-message
-  - bitswap-net
-  - bitswap-wantlist
+    message Entry {
+      bytes block = 1; // CID of the block
+      int32 priority = 2;	// the priority (normalized). default to 1
+      bool cancel = 3; // whether this revokes an entry
+      WantType wantType = 4; // Note: defaults to enum 0, ie Block
+      bool sendDontHave = 5; // Note: defaults to false
+		}
+
+    repeated Entry entries = 1;	// a list of wantlist entries
+    bool full = 2; // whether this is the full wantlist. default to false
+  }
+  message Block {
+    bytes prefix = 1; // CID prefix (all of the CID components except for the digest of the multihash)
```

Comment:
The previous comment here seemed to indicate that this had to be a CIDv1 when it could be a CIDv0 as well.